File: backend/src/services/git.service.ts
import simpleGit, { SimpleGit } from 'simple-git'
import { promises as fs } from 'fs'
import path from 'path'
import { AppError } from '../middleware/errorHandler.js'

const UPLOAD_DIR = process.env.UPLOAD_DIR || './uploads'

export class GitService {
  private git: SimpleGit

  constructor() {
    this.git = simpleGit()
  }

  async cloneRepository(repoUrl: string, projectId: string): Promise<string> {
    try {
      const localPath = path.join(UPLOAD_DIR, projectId)

      // Ensure upload directory exists
      await fs.mkdir(UPLOAD_DIR, { recursive: true })

      // Clone repository
      await this.git.clone(repoUrl, localPath, ['--depth', '1'])

      return localPath
    } catch (error: any) {
      console.error('Git clone error:', error)
      throw new AppError(400, `Failed to clone repository: ${error.message}`)
    }
  }

  async getFileTree(localPath: string): Promise<FileTreeNode> {
    try {
      const name = path.basename(localPath)
      const stats = await fs.stat(localPath)

      if (stats.isDirectory()) {
        const children: FileTreeNode[] = []
        const entries = await fs.readdir(localPath)

        for (const entry of entries) {
          // Skip hidden files and node_modules
          if (entry.startsWith('.') || entry === 'node_modules') {
            continue
          }

          const entryPath = path.join(localPath, entry)
          const child = await this.getFileTree(entryPath)
          children.push(child)
        }

        return {
          path: localPath,
          name,
          type: 'directory',
          children: children.sort((a, b) => {
            // Directories first, then alphabetical
            if (a.type === 'directory' && b.type !== 'directory') return -1
            if (a.type !== 'directory' && b.type === 'directory') return 1
            return a.name.localeCompare(b.name)
          }),
        }
      } else {
        return {
          path: localPath,
          name,
          type: 'file',
          size: stats.size,
          language: this.getLanguageFromPath(localPath),
        }
      }
    } catch (error: any) {
      console.error('File tree error:', error)
      throw new AppError(500, `Failed to read file tree: ${error.message}`)
    }
  }

  async getFileContent(filePath: string): Promise<string> {
    try {
      const content = await fs.readFile(filePath, 'utf-8')
      return content
    } catch (error: any) {
      console.error('File read error:', error)
      throw new AppError(404, `File not found: ${error.message}`)
    }
  }

  async countFiles(localPath: string): Promise<number> {
    let count = 0

    async function countRecursive(dirPath: string) {
      const entries = await fs.readdir(dirPath)

      for (const entry of entries) {
        if (entry.startsWith('.') || entry === 'node_modules') {
          continue
        }

        const entryPath = path.join(dirPath, entry)
        const stats = await fs.stat(entryPath)

        if (stats.isDirectory()) {
          await countRecursive(entryPath)
        } else {
          count++
        }
      }
    }

    await countRecursive(localPath)
    return count
  }

  async getAllCodeFiles(localPath: string): Promise<string[]> {
    const files: string[] = []
    const codeExtensions = [
      '.js', '.jsx', '.ts', '.tsx', '.py', '.java', '.cpp', '.c',
      '.cs', '.go', '.rs', '.rb', '.php', '.swift', '.kt', '.scala',
    ]

    async function scanRecursive(dirPath: string) {
      const entries = await fs.readdir(dirPath)

      for (const entry of entries) {
        if (entry.startsWith('.') || entry === 'node_modules') {
          continue
        }

        const entryPath = path.join(dirPath, entry)
        const stats = await fs.stat(entryPath)

        if (stats.isDirectory()) {
          await scanRecursive(entryPath)
        } else {
          const ext = path.extname(entry)
          if (codeExtensions.includes(ext)) {
            files.push(entryPath)
          }
        }
      }
    }

    await scanRecursive(localPath)
    return files
  }

  private getLanguageFromPath(filePath: string): string {
    const ext = path.extname(filePath).toLowerCase()
    const langMap: Record<string, string> = {
      '.js': 'javascript',
      '.jsx': 'javascript',
      '.ts': 'typescript',
      '.tsx': 'typescript',
      '.py': 'python',
      '.java': 'java',
      '.cpp': 'cpp',
      '.c': 'c',
      '.cs': 'csharp',
      '.go': 'go',
      '.rs': 'rust',
      '.rb': 'ruby',
      '.php': 'php',
      '.swift': 'swift',
      '.kt': 'kotlin',
      '.scala': 'scala',
    }
    return langMap[ext] || 'plaintext'
  }

  async deleteProject(localPath: string): Promise<void> {
    try {
      await fs.rm(localPath, { recursive: true, force: true })
    } catch (error: any) {
      console.error('Delete project error:', error)
    }
  }
}

export interface FileTreeNode {
  path: string
  name: string
  type: 'file' | 'directory'
  children?: FileTreeNode[]
  size?: number
  language?: string
}

File: backend-flask/services/git_service.py
import git
import os
import shutil
from pathlib import Path

class GitService:
    def __init__(self):
        self.upload_dir = os.getenv('UPLOAD_DIR', './uploads')
        os.makedirs(self.upload_dir, exist_ok=True)

    def clone_repository(self, repo_url, project_id):
        """Clone a git repository"""
        local_path = os.path.join(self.upload_dir, project_id)

        try:
            git.Repo.clone_from(repo_url, local_path, depth=1)
            return local_path
        except Exception as e:
            raise Exception(f'Failed to clone repository: {str(e)}')

    def get_file_tree(self, local_path):
        """Get file tree structure"""
        def build_tree(path):
            name = os.path.basename(path)

            if os.path.isdir(path):
                children = []
                try:
                    entries = os.listdir(path)

                    for entry in entries:
                        # Skip hidden files and node_modules
                        if entry.startswith('.') or entry == 'node_modules':
                            continue

                        entry_path = os.path.join(path, entry)
                        child = build_tree(entry_path)
                        if child:
                            children.append(child)

                    # Sort: directories first, then alphabetically
                    children.sort(key=lambda x: (x['type'] != 'directory', x['name']))

                    return {
                        'path': path,
                        'name': name,
                        'type': 'directory',
                        'children': children
                    }

                except PermissionError:
                    return None
            else:
                return {
                    'path': path,
                    'name': name,
                    'type': 'file',
                    'size': os.path.getsize(path),
                    'language': self.get_language_from_path(path)
                }

        return build_tree(local_path)

    def get_file_content(self, file_path):
        """Read file content"""
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                return f.read()
        except Exception as e:
            raise Exception(f'Failed to read file: {str(e)}')

    def get_language_from_path(self, file_path):
        """Determine language from file extension"""
        ext = Path(file_path).suffix.lower()

        lang_map = {
            '.js': 'javascript',
            '.jsx': 'javascript',
            '.ts': 'typescript',
            '.tsx': 'typescript',
            '.py': 'python',
            '.java': 'java',
            '.cpp': 'cpp',
            '.c': 'c',
            '.cs': 'csharp',
            '.go': 'go',
            '.rs': 'rust',
            '.rb': 'ruby',
            '.php': 'php',
            '.swift': 'swift',
            '.kt': 'kotlin',
            '.scala': 'scala',
        }

        return lang_map.get(ext, 'plaintext')

    def delete_project(self, local_path):
        """Delete project directory"""
        try:
            if os.path.exists(local_path):
                shutil.rmtree(local_path)
        except Exception as e:
            print(f'Failed to delete project: {e}')

    def get_all_code_files(self, local_path):
        """Get all code files in project"""
        code_extensions = [
            '.js', '.jsx', '.ts', '.tsx', '.py', '.java', '.cpp', '.c',
            '.cs', '.go', '.rs', '.rb', '.php', '.swift', '.kt', '.scala',
        ]

        code_files = []

        for root, dirs, files in os.walk(local_path):
            # Skip hidden directories and node_modules
            dirs[:] = [d for d in dirs if not d.startswith('.') and d != 'node_modules']

            for file in files:
                if Path(file).suffix.lower() in code_extensions:
                    code_files.append(os.path.join(root, file))

        return code_files
